Fix Conditional Typing in useSWRMutation to Allow Optional ExtraArg Without Explicitly Passing Undefined

saengmotmi opened the pull request:
This PR contains the fixes for the issue: https://github.com/vercel/swr/issues/2665.

This is my first contribution to SWR. Thank you in advance :)

promer94 commented on mutation/types.ts:
```diff
@@ -56,7 +56,7 @@ interface TriggerWithArgs<
 > {
   <SWRData = Data>(
-    extraArgument: ExtraArg,
+    extraArgument?: ExtraArg,
```
I don't think this is the right fix. And it breaks the type validation for required `args`

What we expect here are 

For required `arg`

```tsx
const { trigger, reset, data, error } = useSWRMutation(
    '/api/user',
    (_, opts: { arg: number }) => String(opts.arg)
)
// should throw error, but your change breaks this.
trigger()
// pass type validation
trigger(1)
```

For optional `arg`

```tsx
const { trigger, reset, data, error } = useSWRMutation(
    '/api/user',
    (_, opts: { arg?: number }) => String(opts.arg)
)
// pass type validation
trigger()
// pass type validation
trigger(1)
```

promer94 commented on test/type/trigger.ts:
```diff
@@ -94,9 +94,6 @@ export function useTestSWRMutation() {
   )
 
   const test = () => {
-    // @ts-expect-error `arg` should be 'foo'
```
This part is actually relevant, you shouldn't remove them

promer94 reviewed (changes requested):
I think the correct way to fix this problem is to have a new `TriggerWithOptionsArgs` type. 

So you can improve the type inference implementation here.
https://github.com/vercel/swr/blob/7badbb9d428d2ac48873b3a3f11d54e7cec0c7cf/mutation/types.ts#L142-L144

`ExtraArg = never`  ->  `TriggerWithoutArgs`
`ExtraArg = number | undefined` -> `TriggerWithOptionsArgs`
`ExtraArg = number` -> `TriggerWithArgs`


Also. You need to add a new test for optional `arg`

saengmotmi commented:
@promer94 Oh, I see. I had made incorrect changes. I understand your suggestion. I've made changes according to your suggestion.

I introduced a new `TriggerWithOptionsArgs` type and made the appropriate adjustments for type inference. I have also added a new test for the optional argument.

Please review my latest changes and let me know if there's anything else I need to address.

promer94 reviewed (approved):
Thanks!
